src/main/java/org/jabref/gui/externalfiles/AutoSetFileLinksUtil.java

```diff
@@ -127,16 +138,56 @@ public List<LinkedFile> findAssociatedNotLinkedFiles(BibEntry entry) throws IOEx
 
             if (!fileAlreadyLinked) {
                 Optional<ExternalFileType> type = FileUtil.getFileExtension(foundFile)
-                                                            .map(extension -> ExternalFileTypes.getExternalFileTypeByExt(extension, filePreferences))
-                                                            .orElse(Optional.of(new UnknownExternalFileType("")));
+                                                          .map(extension -> ExternalFileTypes.getExternalFileTypeByExt(extension, filePreferences))
+                                                          .orElse(Optional.of(new UnknownExternalFileType("")));
 
                 String strType = type.isPresent() ? type.get().getName() : "";
                 Path relativeFilePath = FileUtil.relativize(foundFile, directories);
                 LinkedFile linkedFile = new LinkedFile("", relativeFilePath, strType);
                 linkedFiles.add(linkedFile);
             }
         }
-
         return linkedFiles;
     }
+
+    public RelinkedResults relinkingFiles(List<LinkedFile> listlinked) {
+        List<LinkedFile> changedFiles = new ArrayList<>();
+        List<IOException> exceptions = new ArrayList<>();
+
+        for (LinkedFile file : listlinked) {
+            Path path = Path.of(file.getLink());
+            if (!Files.exists(path)) {
+                Path filePath = Path.of(file.getLink());
+                String directoryPath = filePath.getParent().getParent().toString();
+
+                String fileNameString = filePath.getFileName().toString();
+
+                List<String> fileLocations = searchFileInDirectoryAndSubdirectories(Path.of(directoryPath), fileNameString);
+                if (!fileLocations.isEmpty()) {
+                    file.setLink(fileLocations.get(0));
+                    changedFiles.add(file);
+                } else {
+                    exceptions.add(new IOException("couldn't find file: " + file.getLink()));
+                }
+            }
+        }
+        return new RelinkedResults(changedFiles, exceptions);
+    }
+
+    public List<String> searchFileInDirectoryAndSubdirectories(Path directory, String targetFileName) {
+        List<Path> paths = new ArrayList<>();
+        try {
+            Files.walk(directory, Integer.MAX_VALUE, FileVisitOption.FOLLOW_LINKS)
+                 .filter(path -> Files.isRegularFile(path))
+                 .filter(path -> path.getFileName().toString().equals(targetFileName))
+                 .forEach(paths::add);
+        } catch (IOException e) {
+            // Handle any exceptions here
+        }
+        List<String> output = new ArrayList<>();
+        for (Path p : paths) {
+            output.add(p.toString());
```

Comment:
`String` datatype should be used as last resort. Please return `List<Path>` and convert to string at the very last possibility. This enables strong typing.